### PR TITLE
Fixed #31522 -- Made admin's SelectBox preserve scroll position.

### PR DESCRIPTION
**Pull request by TapanGujjar**

### CODE REVIEW

**Comment by TapanGujjar** [comment]
Hi I need help with this PR. The error is couldnot find stderr: fatal: Couldn't find remote ref refs/pull/12833/merge . I have no idea how to solve the error. Thank you

**Comment by felixxm** [comment]
@TapanGujjar Failures are related with force-push. You can rebase your branch and push again.

**Comment by TapanGujjar** [comment]
Hi @felixxm , I rebase the branch and push again. I am getting the error as 
ERROR: Error fetching remote repo 'origin'
hudson.plugins.git.GitException: Failed to fetch from git://github.com/django/django.git

**Comment by felixxm** [comment]
Your branch has conflict with `SelectBox.js` so you didn't rebase from master.

**Comment by TapanGujjar** [comment]
@felixxm thank you for your help. The checks have all been passed. Thank you so much

File: django/contrib/admin/static/admin/js/SelectBox.js [hunk 1]
```diff
@@ -16,17 +16,24 @@
         redisplay: function(id) {
             // Repopulate HTML select box from cache
             const box = document.getElementById(id);
+            const scroll_value_from_top = box.scrollTop;
+            let node;
```
Comment (jdufresne) [inline django/contrib/admin/static/admin/js/SelectBox.js:20]: Why was this moved outside of the `for` block? Can it be moved back?

File: package-lock.json [hunk 2]
```diff
@@ -0,0 +1,3306 @@
+{
```
Comment (jdufresne) [inline package-lock.json:1]: Can you drop the lock file from the commit, please?

File: js_tests/admin/SelectBox.test.js [hunk 3]
```diff
@@ -21,3 +21,38 @@ QUnit.test('filter', function(assert) {
...
+    let anOption;
+    for (let i = 0;i < optionsCount;i++) {
+        anOption = new Option((i + 1), (i + 1), false, false);
+        anOption.title = (i + 1);
+        fromSelectBoxDom.appendChild(anOption);
+    }
+    SelectBox.init(fromId);
+    SelectBox.init(toId);
+
+    for(let i = 0;i < selectedOptionsIndicies.length;i++) {
```
Comment (jdufresne) [inline js_tests/admin/SelectBox.test.js:48]: `for` is a keyword statement, not a function, so space after keyword.

This could iterate using `for ... of`. See #12866.

File: js_tests/admin/SelectBox.test.js [hunk 4]
```diff
@@ -21,3 +21,38 @@ QUnit.test('filter', function(assert) {
...
+
+    const optionsCount = 200;
+    $('<select id="' + fromId + '" multiple ></select>').appendTo('#qunit-fixture');
+    $('<select id="' + toId + '" multiple ></select>').appendTo('#qunit-fixture');
+
+    const fromSelectBoxDom = document.getElementById(fromId);
+    const toSelectBoxDom = document.getElementById(toId);
+    const selectedOptionsIndicies = [194, 195, 196, 197, 198, 199];
+    let anOption;
+    for (let i = 0;i < optionsCount;i++) {
```
Comment (jdufresne) [inline js_tests/admin/SelectBox.test.js:40]:
```suggestion
    for (let i = 0; i < optionsCount; i++) {
```

File: js_tests/admin/SelectBox.test.js [hunk 5]
```diff
@@ -21,3 +21,38 @@ QUnit.test('filter', function(assert) {
...
+    const toId = "to_id";
+
+    const optionsCount = 200;
+    $('<select id="' + fromId + '" multiple ></select>').appendTo('#qunit-fixture');
+    $('<select id="' + toId + '" multiple ></select>').appendTo('#qunit-fixture');
+
+    const fromSelectBoxDom = document.getElementById(fromId);
+    const toSelectBoxDom = document.getElementById(toId);
+    const selectedOptionsIndicies = [194, 195, 196, 197, 198, 199];
+    let anOption;
```
Comment (jdufresne) [inline js_tests/admin/SelectBox.test.js:39]: This is only used inside the `for` block, so can move it there.

File: django/contrib/admin/static/admin/js/SelectBox.js [hunk 6]
```diff
@@ -16,17 +16,24 @@
...
                 if (node.displayed) {
                     const new_option = new Option(node.text, node.value, false, false);
                     // Shows a tooltip when hovering over the option
                     new_option.title = node.text;
                     box.appendChild(new_option);
                 }
             }
+            // reset the scroll bar to the original value
+            box.scrollTop = scroll_value_from_top;
+
```
Comment (jdufresne) [inline django/contrib/admin/static/admin/js/SelectBox.js:36]: Can chop the trailing whitespace.

File: django/contrib/admin/static/admin/js/SelectBox.js [hunk 7]
```diff
@@ -16,17 +16,24 @@
         redisplay: function(id) {
             // Repopulate HTML select box from cache
             const box = document.getElementById(id);
+            const scroll_value_from_top = box.scrollTop;
+            let node;
             box.innerHTML = '';
             const cache = SelectBox.cache[id];
+
+
```
Comment (jdufresne) [inline django/contrib/admin/static/admin/js/SelectBox.js:24]: Can revert this unrelated whitespace change.

**Comment by TapanGujjar** [comment]
Hi @jdufresne, I have applied the changes mentioned by you. Please have a look at it. Thank you

**Comment by felixxm** [comment]
@TapanGujjar Thanks :+1: Welcome aboard :boat: 

I simplified tests.

**Comment by felixxm** [comment]
buildbot, test on selenium.
